app/code/community/Bolt/Boltpay/Model/BoltOrder.php

```diff
@@ -345,6 +351,87 @@ protected function getCorrectedTotal($projectedTotal, $magentoDerivedCartData)
         return $magentoDerivedCartData;
     }
 
+    /**
+     * Checks if billing address of a quote has all the expected fields.
+     * If not, and therefore invalid, the address is replaced by using the
+     * provided shipping address.
+     *
+     * @param Mage_Sales_Model_Quote_Address $billingAddress   The billing address
+     *                                                         to be checked
+     * @param Mage_Sales_Model_Quote_Address $fallbackAddress  The address to fallback
+     *                                                         to in case of invalid billing
+     *                                                         address
+     *
+     *
+     * @return bool  true if a correction is made, otherwise false
+     */
+    public function correctBillingAddress(&$billingAddress, $fallbackAddress)
+    {
+        $quote = $billingAddress->getQuote();
+
+        $wasCorrected = false;
+
+        /////////////////////////////////////////////
+        /// Changes persisted to Magento and Bolt
+        /////////////////////////////////////////////
+        if ($fallbackAddress) {
+            if (
+                !trim($billingAddress->getStreetFull()) ||
+                !$billingAddress->getCity() ||
+                !$billingAddress->getCountry()
+            )
+            {
+                $billingAddress
+                    ->setCity($fallbackAddress->getCity())
+                    ->setRegion($fallbackAddress->getRegion())
+                    ->setPostcode($fallbackAddress->getPostcode())
+                    ->setCountryId($fallbackAddress->getCountryId())
+                    ->setStreet($fallbackAddress->getStreet())
+                    ->save();
+
+                $wasCorrected = true;
+            }
+
+            if (!trim($billingAddress->getName())) {
+                $billingAddress
+                    ->setPrefix($fallbackAddress->getPrefix())
+                    ->setFirstname($fallbackAddress->getFirstname())
+                    ->setMiddlename($fallbackAddress->getMiddlename())
+                    ->setLastname($fallbackAddress->getLastname())
+                    ->setSuffix($fallbackAddress->getSuffix())
+                    ->save();
+
+                $wasCorrected = true;
+            }
+        }
+        if ($wasCorrected) $quote->save();
+        /////////////////////////////////////////////
+
+        /////////////////////////////////////////////
+        /// Changes persisted only to Bolt
+        /// These are workarounds that deal with the
+        /// Bolt requirements for fields that are not
+        /// required by Magento and should not be
+        /// persisted in Magento
```

Comment:
Mutating address object but not saving to magento seems very weird to me. If some code in code path calls $address->save later, these change will be written to database.

Given we only want to change data for $cartSubmissionData, let's set fallback in where we construct $cartSubmissionData ?

Reply:
@daisy1754 makes sense.